### DISCUSSION TITLE
PLEASE CLOSE THIS ONE

### DISCUSSION THREAD

**Discussion by dorpxam** [AkaneTendo25/musubi-tuner#9] [Ideas]
> This does not appear to be viable, as the [LTX-2 source code](https://github.com/Lightricks/LTX-2/blob/main/LICENSE), upon which this implementation is built, is not compatible with Apache 2.0, used by musubi-tuner.

After working on the code for the console colorization, I see that there's a lot of intricate from the `musubi-tuner` engine. Maybe, it's one of the problem of the current implementation and the introduction of regression. 

Understand me correctly, this isn't a criticism of your work, which I still find excellent. But as someone who originally developed C++, I'm aware of the implications of intricate code. I don't have the technical expertise to help you with complex tasks involving deep learning concepts. But I can help you with everything else.

Let's state the problem pragmatically:

1. This project will not be a pull request for musubi-tuner given the use of code portions from Lightricks' LTX-2 trainer
2. The current code relies on musubi-tuner by intertwining Lightricks code with your own intermediate code.
3. The main idea is to have a LoRa training code (and potentially fine-tuning) for the LTX-2 model (and only LTX-2 for now) that is fast and usable with common GPUs (consumer-grade GPUs). Potentially with low VRAM.

It might be wise to detach ourselves from the musubi-tuner code and extract only the useful elements. For example, like SimpleTuner, which uses a similar block-swap technique to musubi-tuner.

As a developer, I know it's much simpler to isolate code by class (whether object-oriented or functional). And I know it's much simpler to maintain your own code rather than scribbling over someone else's.

For example, the code I introduced for console colorization is fairly well isolated, except for the logging system, whose text is directly included in the code, which poses a real maintenance problem. Ideally, and this would also allow for internationalization of the text, the log text should be isolated in a JSON file or similar format and accessed via an identifier (e.g., a unique identifier: UUID). This would allow for the logging to be displayed with or without colorization, and for switching between the two. Do you see the idea?

Furthermore, regarding the current code regression issue, this would allow for a much better approach in terms of code and performance benchmarking using appropriate tools. I'm certain we can achieve sufficiently abstract code that can be extended with future versions or new models.

I know it's a big undertaking, but if we can isolate the code from musubi-tuner's, I'm sure maintenance will be simpler. Let's keep simplicity in mind; there's no need for code as cumbersome as SimpleTuner's. The goal is to have a usable and robust version in the coming days.

What do you think?

**Comment by AkaneTendo25** [comment]
I understand your points. LTX-2 implementation was initially meant to be a PR for musubi-tuner, so I never planned to turn it into a standalone project. Otherwise, I'd probably build it on top of the original trainer, using either RamTorch/TorchAO or simply reusing musubi's block swapping and scaled quantization logic (those are pretty well encapsulated and documented, so they should be easy to reuse in another project, and musubi's license is very generous).  

But I do plan to build my own standalone trainer eventually, so thanks for discussing these points. I'll keep your ideas in mind. And yeah, I dont't have a strong ML background either, otherwise I probably wouldn't have spent a month implementing LTX-2 LoRA training 😅 (honestly, I expected it to be easier than Kandinsky 5). While I have 10+ years of software development experience (mostly web/full-stack), I've only been working (professionally) as an AI video engineer since December, so I’m still building up my ML/DL skills. That's also why I'll definitely keep making trainers for various upcoming video (and music) models.

**Comment by dorpxam** [comment]
> And yeah, I dont't have a strong ML background either, otherwise I probably wouldn't have spent a month implementing LTX-2 LoRA training

Don't be so modest, it's already a huge job, especially combining existing code from diverse sources (LTX-2, musubi-tuner and more)

I wish I knew as much. Even though I understand some of the basics, it's too complex for me.

However, I don't know why, but I have a strong feeling that the legacy branch shouldn't have been scrapped. There wasn't a bug per se, just a convergence issue during training. It's possible your optimizations were perfect, but you might have introduced a tiny, insignificant bug that you couldn't fix.

The line between convergence and non-convergence might depend on a simple, trivial parameter that was changed. That's why I thought the idea of ​​starting from the working commit and working backward to the failing one would have allowed us to isolate the problem, or at least narrow down the search.

I may not be clairvoyant, or even a fortune teller, but I have a good instinct for these things. 😄

**Comment by dorpxam** [comment]
I think this discussion is close now 😃 

I see you've open start 'Fine Tuning' code, what a good news! I hope that will be possible locally. Seriously, I love LoRA and now we can jump to the Rank-64 bath. But what a hell for the size (600MB in Rank-32). Strategically, I think that can be a big improvement for the community.

It's clear that the LTX-2 model it's trained on low end videos. A lot of Indian content for sure, probably for a copyright question and law suit issue. Good fine-tuning will produce very interesting result I'm sure. We can already get a glimpse of this with LoRa!
